### ISSUE TITLE
[python] adding methods to nodes fails with alternatives

### ISSUE THREAD

**Issue by epavese** [jcoglan/canopy#30]
Basing this on the example grammar

```
grammar Choice
  root  <-  "a" "b" "c" <Foo> / "1" "2" "3" <Bar>

```
no matter what Foo and Bar define, this will fail on input "123" with:
```
  File "grammar.py", line 95, in _read_root
    address0.__class__ = type(cls0.__name__ + 'Foo', (cls0, self._types.Foo), {})
TypeError: Cannot create a consistent method resolution
order (MRO) for bases Foo, object

```
Looking at the generated grammar the problem seems to be that it is trying to set address.__class__ *before* checking for FAILURE on the match (in this case the match on "abc" fails, a Foo node should not be created, but it is trying to assign that type. Testing for FAILURE before attempting to set the class seems to work

**Comment by vitenbergd** [comment]
Hello, first of all, @jcoglan  thanks a lot for canopy. I've found the same issue when i tried to use https://github.com/imjoey/pyhaproxy project, and @epavese is absolutely right - it happens in situations when **address0** is **FAILURE**.

This is similar to:
```python
>>> class A(object): pass
... 
>>> type('B', (object,A), {})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Cannot create a consistent method resolution
order (MRO) for bases object, A
```
In the generated code it looks like:
```python
        if elements0 is None:
            address0 = FAILURE
        else:
            address0 = TreeNode1(self._input[index1:self._offset], index1, elements0)
            self._offset = self._offset
        cls0 = type(address0)
        address0.__class__ = type(cls0.__name__ + 'Global', (cls0, self._types.Global), {})
```
So i think it's indentation of the generated code issue, cause last 2 lines imho belong to the **else** branch of this code, so the code below worked for me:
```python
        if elements0 is None:
            address0 = FAILURE
        else:
            address0 = TreeNode1(self._input[index1:self._offset], index1, elements0)
            self._offset = self._offset
            cls0 = type(address0)
            address0.__class__ = type(cls0.__name__ + 'Global', (cls0, self._types.Global), {})
```

**Comment by jcoglan** [comment]
@vitenbergd I don't believe your fix will work in general, because it relies makes assumptions about the context in which `extendNode_()` is applied by using `_indent()` to turn this code into a continuation of whatever the previous code block is.

One place where this won't work is in the following grammar:

```
grammar Any

any <- . <Foo>
```

This will generate code looking something like this:

```py
class Grammar(object):
    def _read_any(self):
        # ...
        if self._offset < self._input_size:
            address0 = TreeNode(self._input[self._offset:self._offset + 1], self._offset, [])
            self._offset = self._offset + 1
        else:
            address0 = FAILURE
            # ...
        cls0 = type(address0)
        address0.__class__ = type(cls0.__name__ + 'Foo', (cls0, self._types.Foo), {})
        self._cache['any'][index0] = (address0, self._offset)
        return address0
```

If we apply your fix then we get the `cls0 = ...` and the following line indented:

```py
class Grammar(object):
    def _read_any(self):
        # ...
        if self._offset < self._input_size:
            address0 = TreeNode(self._input[self._offset:self._offset + 1], self._offset, [])
            self._offset = self._offset + 1
        else:
            address0 = FAILURE
            # ...
            cls0 = type(address0)
            address0.__class__ = type(cls0.__name__ + 'Foo', (cls0, self._types.Foo), {})
        self._cache['any'][index0] = (address0, self._offset)
        return address0
```

They thus become part of the failure branch. So essentially this code breaks a core assumption in the code generator which is that individual blocks of code are independent.

I'm now examining #39 and may apply that or my own patch.

**Comment by jcoglan** [comment]
This has now been fixed in a way I believe is generally workable in 5081a30. Could one of you check this and confirm, please?

**Comment by jcoglan** [comment]
I'm closing this as @longhotsummer has confirmed the fix. Is someone else finds that this is still a problem, please leave further comments here.

**Comment by jcoglan** [comment]
This should now be fixed as part of the v0.4 release; see https://blog.jcoglan.com/2022/04/01/announcing-canopy-0-4/
